reconcile/endpoints_discovery/integration.py

```diff
@@ -0,0 +1,356 @@
+
+
+class EndpointsDiscoveryIntegration(
+    QontractReconcileIntegration[EndpointsDiscoveryIntegrationParams]
+):
+    """Discover routes from all OpenShift clusters and update endPoints in app-interface."""
+
+    @property
+    def name(self) -> str:
+        return QONTRACT_INTEGRATION
+
+    def get_early_exit_desired_state(
+        self, query_func: Callable | None = None
+    ) -> dict[str, Any]:
+        """Return the desired state for early exit."""
+        if not query_func:
+            query_func = gql.get_api().query
+        return {
+            "namespaces": [ns.dict() for ns in self.get_namespaces(query_func)],
+        }
+
+    def get_desired_state_shard_config(self) -> None:
+        """Sharding (per cluster) is not supported for this integration.
+
+        An application can have endpoints in multiple clusters and this may cause merge conflicts."""
+        return None
+
+    def get_namespaces(
+        self,
+        query_func: Callable,
+        cluster_names: Iterable[str] | None = None,
+        namespace_name: str | None = None,
+    ) -> list[NamespaceV1]:
+        """Return namespaces to consider for the integration."""
+        return [
+            ns
+            for ns in namespaces_query(query_func).namespaces or []
+            if integration_is_enabled(self.name, ns.cluster)
+            and (not cluster_names or ns.cluster.name in cluster_names)
+            and (not namespace_name or ns.name == namespace_name)
+            and not ns.delete
+        ]
+
+    def get_routes(self, oc_map: OCMap, namespace: NamespaceV1) -> list[Route]:
+        """Return the routes for the given namespace."""
+        oc = oc_map.get_cluster(namespace.cluster.name)
+        if not oc.project_exists(namespace.name):
+            logging.info(
+                f"{namespace.cluster.name}/{namespace.name}: Namespace does not exist (yet). Skipping for now!"
+            )
+            return []
+
+        return [
+            Route(
+                name=item["metadata"]["name"],
+                host=item["spec"]["host"],
+                tls=bool(item["spec"].get("tls")),
+            )
+            for item in oc.get_items(kind="Route", namespace=namespace.name)
+        ]
+
+    def get_endpoint_changes(
+        self,
+        app: str,
+        endpoint_prefix: str,
+        endpoint_template: str,
+        endpoints: Iterable[AppEndPointsV1],
+        routes: Iterable[Route],
+    ) -> tuple[EndpointsToAdd, EndpointsToChange, EndpointsToDelete]:
+        """Get all new/changed/deleted endpoints for the given namespace."""
+        if not routes and not endpoints:
+            # nothing to do
+            return [], [], []
+
+        diff = diff_any_iterables(
+            current=endpoints or [],
+            desired=routes,
+            # use the hostname as the key for both current and desired
+            current_key=lambda endpoint: extract_hostname(endpoint.url),
+            desired_key=lambda route: route.host,
```

Comment:
Mhhh, good question! First, I thought, this couldn't be, because the integration doesn't add a new endpoint if there is already an endpoint with the hostname. BUT! A user could add a new endpoint manually afterward! We shouldn't support that! I'll throw an exception in this case.

Reply:
a19c243 refactors that. The integration adds and manages `app.endPoints` for all discovered routes.